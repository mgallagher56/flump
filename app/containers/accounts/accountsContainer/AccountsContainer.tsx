import type { FC } from 'react';
import { Fragment, useMemo } from 'react';

import { useLoaderData } from '@remix-run/react';
import { useTranslation } from 'react-i18next';
import AccountsCard from '~/components/core/cards/AccountsCard';
import FLPBox from '~/components/core/structure/FLPBox';
import FLPHeading from '~/components/core/typography/FLPHeading';
import type { loader } from '~/routes/app.accounts._index';

import AddEditAccountsDialogBtn from '../../../components/dialogs/addEditAccountsDialog.tsx/AddEditAccountsDialog';
import { AccountTypeEnum, isAccountTypeValid } from '../utils';

const { CURRENT, SAVING, MORTGAGE, CREDIT_CARD, LOAN, OWED } = AccountTypeEnum;

const AccountsContainer: FC = () => {
  const { t } = useTranslation();
  const { accounts } = useLoaderData<typeof loader>();
  const accountTypeArr = useMemo(() => [CURRENT, SAVING, CREDIT_CARD, MORTGAGE, LOAN, OWED], []);

  return (
    <>
      <FLPBox mb={5} display="flex" flexDirection="row" justifyContent="space-between">
        <FLPHeading size="xl" as="h1">
          {t('accounts')}
        </FLPHeading>
        <AddEditAccountsDialogBtn />
      </FLPBox>
      {accountTypeArr.map((accountType) => {
        return (
          <Fragment key={accountType}>
            {isAccountTypeValid(accountType, accounts) && (
              <FLPHeading as="h2" size="lg">
                {accountType}
              </FLPHeading>
            )}
            <FLPBox display="flex" flexWrap="wrap" gap={5}>
              {accounts.map(
                ({ name, id, type }: { name?: string; id?: string; type?: AccountTypeEnum | string }) =>
                  type === accountType && <AccountsCard key={id} accountId={id} name={name} type={type} />
              )}
            </FLPBox>
          </Fragment>
        );
      })}
    </>
  );
};

export default AccountsContainer;
