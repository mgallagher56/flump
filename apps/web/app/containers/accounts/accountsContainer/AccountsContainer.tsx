import type { FC } from 'react';
import { Fragment, useMemo } from 'react';

import { useTranslation } from 'react-i18next';
import { useLoaderData } from 'react-router';
import AccountsCard from '~/components/core/cards/AccountsCard';
import FLPBox from '~/components/core/structure/FLPBox';
import FLPHeading from '~/components/core/typography/FLPHeading';
import AddEditAccountsDialogBtn from '~/components/dialogs/addEditAccountsDialog.tsx/AddEditAccountsDialog';
import { AccountTypeEnum, isAccountTypeValid } from '~/containers/accounts/utils';
import type { loader } from '~/routes/app.accounts._index';

const { CURRENT, SAVING, MORTGAGE, CREDIT_CARD, LOAN, OWED } = AccountTypeEnum;

const AccountsContainer: FC = () => {
  const { t } = useTranslation();
  const { accounts } = useLoaderData<typeof loader>();
  const accountTypeArr = useMemo(() => [CURRENT, SAVING, CREDIT_CARD, MORTGAGE, LOAN, OWED], []);

  return (
    <>
      <FLPBox display="flex" flexDirection="row" justifyContent="space-between" my={5}>
        <FLPHeading as="h1" size="xl">
          {t('accounts')}
        </FLPHeading>
        <AddEditAccountsDialogBtn />
      </FLPBox>
      <FLPBox display="flex" flexWrap="wrap" gap={10}>
        {accountTypeArr.map((accountType) => {
          return (
            isAccountTypeValid(accountType, accounts) && (
              <Fragment key={accountType}>
                <FLPBox>
                  <FLPHeading as="h2" size="lg">
                    {accountType}
                  </FLPHeading>
                  <FLPBox display="flex" flexWrap="wrap" gap={5}>
                    {accounts.map(
                      ({ name, id, type }: { name?: string; id?: string; type?: AccountTypeEnum | string }) =>
                        type === accountType && <AccountsCard key={id} accountId={id} name={name} type={type} />
                    )}
                  </FLPBox>
                </FLPBox>
              </Fragment>
            )
          );
        })}
      </FLPBox>
    </>
  );
};

export default AccountsContainer;
