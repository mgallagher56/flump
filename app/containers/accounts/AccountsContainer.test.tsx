import type { ReactNode } from 'react';

import { fireEvent, render, waitFor } from '@testing-library/react';
import { vi } from 'vitest';

import mockAccounts from '__mocks__/accounts';

import AccountsContainer from './AccountsContainer';

const mocks = vi.hoisted(() => ({
  mockUseLoaderData: vi.fn()
}));

vi.mock('@remix-run/react', async () => {
  const actual: Record<string, unknown> = await vi.importActual('@remix-run/react');
  return {
    ...actual,
    useLoaderData: mocks.mockUseLoaderData,
    Form: ({ children }: { children: ReactNode }) => <form>{children}</form>,
    useSubmit: () => ({ onSubmit: vi.fn() })
  };
});

vi.mock('app/components/core/cards/AccountsCard', () => ({ default: () => 'AccountsCard' }));
vi.mock('app/components/core/input/FLPInput', () => ({ default: () => 'FLPInput' }));

describe('<AccountsContainer />', () => {
  it('should render as expected when no accounts exist', () => {
    mocks.mockUseLoaderData.mockReturnValue({ accounts: [] });
    const { container } = render(<AccountsContainer />);
    expect(container).toMatchSnapshot();
  });

  it('should render as expected when accounts exist', () => {
    mocks.mockUseLoaderData.mockReturnValue({ accounts: mockAccounts });
    const { container } = render(<AccountsContainer />);
    expect(container).toMatchSnapshot();
  });

  it('should call supabase functions to add an account', async () => {
    const { getByText } = render(<AccountsContainer />);
    const addAccountModalBtn = getByText('addAccount');
    expect(addAccountModalBtn).toBeDefined();
    fireEvent.click(addAccountModalBtn);
    await waitFor(() => {
      expect(getByText('addAccount')).toBeDefined();
    });
  });
});
