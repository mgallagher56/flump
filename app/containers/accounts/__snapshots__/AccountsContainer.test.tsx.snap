// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AccountsContainer /> > should render as expected when accounts exist 1`] = `
<div>
  <div
    class="css-1qtey63"
  >
    <h1
      class="chakra-heading css-1e3tj"
    >
      accounts
    </h1>
    <button
      aria-controls="dialog::r2::content"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="chakra-button css-4xx2wk"
      data-part="trigger"
      data-scope="dialog"
      data-state="closed"
      id="dialog::r2::trigger"
      type="button"
    >
      addAccount
    </button>
  </div>
  <h2
    class="chakra-heading css-1e3tj"
  >
    Current
  </h2>
  <div
    class="css-1terknm"
  >
    AccountsCard
    AccountsCard
  </div>
  <h2
    class="chakra-heading css-1e3tj"
  >
    Saving
  </h2>
  <div
    class="css-1terknm"
  >
    AccountsCard
    AccountsCard
    AccountsCard
  </div>
  <h2
    class="chakra-heading css-1e3tj"
  >
    Mortgage
  </h2>
  <div
    class="css-1terknm"
  >
    AccountsCard
  </div>
  <h2
    class="chakra-heading css-1e3tj"
  >
    Loan
  </h2>
  <div
    class="css-1terknm"
  >
    AccountsCard
  </div>
  <h2
    class="chakra-heading css-1e3tj"
  >
    Owed
  </h2>
  <div
    class="css-1terknm"
  >
    AccountsCard
  </div>
</div>
`;

exports[`<AccountsContainer /> > should render as expected when no accounts exist 1`] = `
<div>
  <div
    class="css-1qtey63"
  >
    <h1
      class="chakra-heading css-1e3tj"
    >
      accounts
    </h1>
    <button
      aria-controls="dialog::r0::content"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="chakra-button css-4xx2wk"
      data-part="trigger"
      data-scope="dialog"
      data-state="closed"
      id="dialog::r0::trigger"
      type="button"
    >
      addAccount
    </button>
  </div>
  <div
    class="css-1terknm"
  />
  <div
    class="css-1terknm"
  />
  <div
    class="css-1terknm"
  />
  <div
    class="css-1terknm"
  />
  <div
    class="css-1terknm"
  />
</div>
`;
